### ISSUE TITLE
AnyFixture vs before_all

### ISSUE THREAD

**Issue by AlexVPopov** [palkan/test-prof#28]
It seems both accomplish quite a similar goal - create shared setup only once instead of per example. However, I'm not sure which one is better to use. What are the considerations, that one has to take into account when choosing? Or are they complementary?

**Comment by AlexVPopov** [comment]
Btw, sorry if it's wrong to post questions here, but there's no **Support** section in the Readme and that's why I chose to write here.

**Comment by palkan** [comment]
`before_all` works only within an example group and should be considered as a replacement for `before(:each)` (when there is a lot of affected examples) and `before(:all)` ('cause it's not safe regarding to DB _cleanliness_).

`AnyFixture`, from the other hand, is kind of replacement for traditional, YAML-based, fixtures. The data generated within `AnyFixture` persists for the whole run, not only for the example group where it has been activated.

Take a look at the recent [slides](https://speakerdeck.com/palkan/railsclub-moscow-2017-faster-tests), where I give some explanations and tips when to use each tool (video is not there yet, unfortunately).

TLDR; Use `AnyFixture` when you have similar setup **between** files and use `before_all/let_it_be` to share setup within an example group.

**Comment by AlexVPopov** [comment]
Thank you very much for the clarification. I use `AnyFixture` heavily in my test suit. While it certainly keeps the speed of the suite high, it has several times bitten me in unexpected ways. When running my specs in random order I am never sure what is in the database. Also, sometimes tests may pass, when run as part of the whole suite, but might fail when running stand-alone. The reason is that, for example, they expect a record, which has been created in another file (full suite run), but is not created in the current file (individual suite run). So one has to be very careful. I guess one has to trade speed against readability and maintainability.

**Comment by palkan** [comment]
Well, yes, the same drawbacks as with usual fixtures. And that's the reason when other tools (such as `before_all`) exist–to avoid global state.

Btw, there is one useful thing missing in the guide–[`:with_clean_fixture`](https://github.com/palkan/test-prof/blob/master/lib/test_prof/recipes/rspec/any_fixture.rb) shared context, which allows you to _disable_ `AnyFixture` data for an example (i.e. allows you to work with clean state):

```ruby
describe User do
  before(:all) { @account = TestProf::AnyFixture.register(:account) { create(:account) } }

  it "assigns new user to the first account if not specified" do
     expect(create(:user).account).to eq @account
  end

  it "raises an error if no accounts", :with_clean_fixture do
    expect { create(:user) }.to raise_error(ActiveRecord::RecordInvalid)
  end
end
```

**Comment by AlexVPopov** [comment]
Wow, this is indeed super useful!!! I'll definitely take it into account!
